samples/xbee/demos/EndToEndAgricultureSample/main.py

```diff
@@ -841,6 +850,36 @@ def upload_configuration_drm(configuration, sender=None):
         print_error("Could not upload datapoint to stream '{}': {}".format(data_stream, str(e)))
 
 
+def upload_configurations_drm(configurations, sender=None):
+    """
+    Uploads the given configurations to Digi Remote Manager.
+
+    Args:
+        configurations (list of tuple): the configurations to upload to DRM.
+        sender (String, optional): the 64-bit address of the solar panel that
+            originated the configuration. Use ``None`` if the configuration is
+            from the solar controller.
+    """
+    # Check if DataPoint module is imported
+    if "DataPoint" not in sys.modules:
+        from digidevice.datapoint import DataPoint
+    data_points = []
+    for config in configurations:
+        conf_id = config[0]
+        conf_value = config[1]
+
+        # Generate the corresponding data stream.
+        data_stream = DATA_STREAM_FORMAT.format(sender, conf_id) if sender is not None else conf_id
+        # Create and store the data point.
+        data_points.append(DataPoint(data_stream, conf_value, data_type=DataType.DOUBLE))
```

Comment:
To do the import we are checking if `DataPoint` is in `sys.modules`, but here we are using it independently of this result.
I guess the use of `DataPoint` depends on the version of the firmware: for 21.8 or higher it will be there but not for lower versions.

Why do not have just one `upload_configurations_drm()` function and distinguish inside the method to use? (Not tested)

```python
def upload_configurations_drm(configurations, sender=None):
    """
    Uploads the given configurations to Digi Remote Manager.

    Args:
        configurations (list of tuple): the configurations to upload to DRM.
        sender (String, optional): the 64-bit address of the solar panel that
            originated the configuration. Use ``None`` if the configuration is
            from the solar controller.
    """
    if is_old_firmware():
        for config in configurations:
            conf_id = config[0]
            conf_value = config[1]

            # Generate the corresponding data stream.
            data_stream = DATA_STREAM_FORMAT.format(sender, conf_id) if sender is not None else conf_id
            # Upload the measurement as a new data point of the data stream.
            try:
                with datapoint_lock:
                    datapoint.upload(data_stream, conf_value, data_type=datapoint.DataType.DOUBLE)
            except Exception as e:
                print_error("Could not upload datapoint to stream '{}': {}".format(data_stream, str(e)))
    else:
        from digidevice.datapoint import DataPoint
        data_points = []
        for config in configurations:
            conf_id = config[0]
            conf_value = config[1]

            # Generate the corresponding data stream.
            data_stream = DATA_STREAM_FORMAT.format(sender, conf_id) if sender is not None else conf_id
            # Create and store the data point.
            data_points.append(DataPoint(data_stream, conf_value, data_type=DataType.DOUBLE))

            # Upload the data points all at once.
            try:
                with datapoint_lock:
                    datapoint.upload_multiple(data_points)
            except Exception as e:
                print_error("Could not upload datapoints: {}".format(str(e)))
```

When `upload_configuration_drm()`(without the `s`) was used, we should create a list and call `upload_configurations_drm()`:

* In `xbee_data_callback(msg)` just call `upload_configurations_drm(configurations)`:
  ```python
    [...]
    # Update the valve position and last moisture in the dictionaries.
    if STAT_VALVE in properties:
        station_valve_dict[sender] = True if properties[STAT_VALVE] == 1 else False
    if STAT_MOISTURE in properties:
        station_moisture_dict[sender] = float(properties[STAT_MOISTURE])

    # Parse the configurations contained in the message and upload them to DRM.
    upload_configurations_drm(parse_configurations(properties), sender)
    [...]
  ```
* The same for `drm_report_tast()`:
  ```python
    [...]
    configurations.append((STAT_RAIN, rain_f))
    configurations.append((STAT_WIND, wind))
    configurations.append((STAT_RADIATION, radiation))
    configurations.append((STAT_LEVEL, tank_level_f))
    configurations.append((STAT_VALVE, int(tank_valve_open)))

    upload_configurations_drm(configurations)

    # Create and start a timer to repeat this task periodically.
    t = Timer(report_interval, drm_report_task)
    [...]
  ```

* Create a list in `set_tank_valve(is_open)`:
  ```python
    [...]
    global tank_valve_open

    tank_valve_open = is_open

    # Upload a datapoint with the new position.
    Thread(target=upload_configurations_drm, args=([(STAT_VALVE, int(is_open))],), daemon=True).start()

    print_log("Tank valve is now {}".format("open" if tank_valve_open else "closed"))
    return AT_VALUE_ENABLED if is_open else AT_VALUE_DISABLED
    [...]
  ```
* Create a list in `refill_tank()`:
  ```python
    [...]
    global tank_level

    tank_level = 100

    # Upload a datapoint with the new level.
    Thread(target=upload_configuration_drm, args=([(STAT_LEVEL, tank_level)],), daemon=True).start()

    print_log("Tank refilled")
    return tank_level
    [...]
  ```